model/src/app_data.rs

```diff
@@ -0,0 +1,134 @@
+//! Contains the app_data file structures to define additional data about tx origin
+
+use crate::h160_hexadecimal::{self};
+use anyhow::{anyhow, Result};
+use cid::multihash::{Code, MultihashDigest};
+use primitive_types::{H160, H256};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use serde_with::serde_as;
+use std::convert::TryInto;
+
+#[serde_as]
+#[derive(Eq, PartialEq, Clone, Debug, Deserialize, Serialize, Hash, Default)]
+pub struct Referrer {
+    #[serde(with = "h160_hexadecimal")]
+    pub address: H160,
+    pub version: String,
+}
+
+#[serde_as]
+#[derive(Eq, PartialEq, Clone, Debug, Deserialize, Serialize, Hash, Default)]
+pub struct Metadata {
+    pub referrer: Option<Referrer>,
+}
+
+#[serde_as]
+#[derive(Eq, PartialEq, Clone, Debug, Deserialize, Serialize, Hash, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct AppData {
+    pub version: String,
+    pub app_code: Option<String>,
+    pub metadata: Option<Metadata>,
+}
+
+#[serde_as]
+#[derive(Eq, PartialEq, Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct AppDataBlob(pub Value);
+
+impl AppDataBlob {
+    pub fn sha_hash(&self) -> Result<H256> {
+        // The following hash is the hash used by ipfs.
+        // The ipfs cid can be calculated by
+        // const RAW: u64 = 0x55;
+        // let hash = Code::Sha2_256.digest(&string.into_bytes());
+        // let cid = Cid::new_v1(RAW, hash);
+        // In order to avoid json duplication, we are deriving the hash from the json object
+        let hash = Code::Sha2_256.digest(serde_json::ser::to_string(&self.0.clone())?.as_bytes());
+        let array: [u8; 32] = hash.to_bytes()[2..]
```

Comment:
ubernit: I found it a bit confusing that we skip the first two bytes of the digest, as in my mental model Sha2-256 digests are 32 bytes.

After some digging, it turns out that its because `digest` returns a `Multihash` and not the digest - so it includes the multihash code (0x12 for Sha2) and the length (as a variable length integer, so here just 1 byte `0x20`) followed by the actual 32-byte digest.


It turns out that `Multihash` has a [`digest`](https://docs.rs/multihash/0.14.0/multihash/struct.MultihashGeneric.html#method.digest) method that extracts just the digest from the multihash and might avoid some confusion:
```suggestion
        let array: [u8; 32] = hash.digest()
```